From 1925a0377980a9a11fc475ee352afa6620da258b Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Thu, 11 Nov 2021 08:11:17 +0200
Subject: [PATCH] [SW-36216] gaudi2: security: add enable check api

Add security enable check api which is a precondition
for secuity initialization.

Change-Id: I25a34198b944899af5958ad183d4ad698a7c84b3
---

diff --git a/arch/gaudi2/arm/common.c b/arch/gaudi2/arm/common.c
index f03b3ae..3d89e9e 100644
--- a/arch/gaudi2/arm/common.c
+++ b/arch/gaudi2/arm/common.c
@@ -42,6 +42,7 @@
 #include "common.h"
 #include <hl.h>
 #include <address_space.h>
+#include <efuse.h>
 
 
 #define ARC0_ICCM_BASE_ADDR		0x1000007FF7000000ull
@@ -405,3 +406,39 @@
 	if (__stack2 != STACK_CANARY_MAGIC_WORD)
 		cpu_boot_err_set(CPU_BOOT_ERR_BOOT_FW_CRIT_ERR);
 }
+
+#define EFUSE_SECURITY_ENABLE_BIT		BIT(15)
+#define EFUSE_SECURITY_WORD_OFFSET		0x3A
+#define BOOT_STRAP_SECURITY_ENABLE_BIT		BIT(0)
+
+/*
+ * The routine shall return true in case of security enable in strap.
+ * Otherwise, shall return false (The polarity of strap is inverted:
+ * meaning '0' is for security enable).
+ */
+static bool hl_is_secuirty_strap_enabled(void)
+{
+	u32 boot_strap;
+
+	boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS_H);
+
+	if (boot_strap & BOOT_STRAP_SECURITY_ENABLE_BIT)
+		return false;
+
+	return true;
+}
+
+bool preboot_security_is_enabled(void)
+{
+	u32 efuse_word;
+
+	/* If security is enabled via boot strap pins */
+	if (hl_is_secuirty_strap_enabled())
+		return true;
+
+	/* If eFuse read failed, assume an attacker sabotage, security in on */
+	if (preboot_fuse_read(EFUSE_SECURITY_WORD_OFFSET, &efuse_word))
+		return true;
+
+	return !!(efuse_word & EFUSE_SECURITY_ENABLE_BIT);
+}
